### ISSUE TITLE
Goto/switch/loop issues and possible codegen changes

### ISSUE THREAD

**Issue by kg** [sq/JSIL#902]
Since code containing gotos (at least as understood by ILSpy) is a consistent problem, and this tends to interact with switches and loops, I'm thinking a stopgap fix might be appropriate.

One idea that occurs to me is that functions could be translated in two 'modes', where one of the modes is 'goto-compatible' and it takes *all* control flow apart (if statements, switch statements, for loops) into labels and gotos. Then the function is just a single label dispatch loop with labels, and the codegen should be more reliable. Anyone see a problem with that?

Alternately I get around to writing a SSA decompiler and we can light ILSpy on fire and throw away the ashes, but I don't think I'll have time for that this decade...

**Comment by iskiselev** [comment]
How would be  'goto-compatible' mode selected? Is it something opted-in with MetaAttribute?
Such change will make readibility of translated source much worse, but it already bad with lambdas and asyn/await. So IMHO, it is not big problem - we can improve source maps to compensate it.

**Comment by kg** [comment]
I think it would just get turned on any time a function contains goto. Goto is very rare. That would make it very easy to identify whether it should be active, and it would simplify all the code involved in handling gotos and control flow.

**Comment by iskiselev** [comment]
Original function in #899 has no goto. Do you mean that function translated by IlSpy contains at least one goto as border condition?

**Comment by kg** [comment]
Yeah, the goto appears because of how ILSpy decompiles the output of csc. Technically msil contains plenty of gotos :-)

**Comment by iskiselev** [comment]
:) msil contains goto if you use any cycle, but it is too strong condition, with it nearly all functions will be translated in go-to friendly mode :)

**Comment by kg** [comment]
Well, we'd be detecting it based on whether ilspy introduced one, so that narrows it down. The way JSIL handles function bodies containing an ILSpy goto right now is really bad, so I specifically want to change that

**Comment by markusjohnsson** [comment]
I'm on the other side here, I've experimented with turning IL into JS myself and use only a `while(...){ switch(pos) {...} }`- construct per block (method/try/catch/finally). It is a very slow pattern.. I would need to generate if/else statements and for loops to make it faster.. but as a last-resort for, as you say, goto-compatible code, it is quite reliable.

Feel free to check the code out: https://github.com/markusjohnsson/cil.js

To see what kind of code it generates, check: https://github.com/markusjohnsson/cil.js/blob/master/CilJs.TestRunner/corlib.ciljs.js#L504
